File: java/src/main/java/io/yeeco/yeesigner/Address.java
package io.yeeco.yeesigner;

import org.javatuples.Pair;

public class Address {

    public static String encode(byte[] publicKey, String hrp) throws SignerException {

        byte[] hrpBytes = hrp.getBytes();

        byte[] error = new byte[1];
        long vecPointer = JNI.addressEncode(publicKey, hrpBytes, error);
        ErrorUtils.checkErrorCode(error[0]);

        byte[] result = Utils.copyAndFreeVec(vecPointer, error);

        String address = null;
        try {
            address = new String(result, "UTF-8");
        } catch (Exception e) {
            throw new SignerException("charset error");
        }

        return address;
    }

    public static Pair<byte[], String> decode(String address) throws SignerException {

        byte[] addressBytes = address.getBytes();

        byte[] error = new byte[1];
        long[] publicKeyVecPointer = new long[1];
        long[] hrpVecPointer = new long[1];
        JNI.addressDecode(addressBytes, publicKeyVecPointer, hrpVecPointer, error);
        ErrorUtils.checkErrorCode(error[0]);

        byte[] publicKey = Utils.copyAndFreeVec(publicKeyVecPointer[0], error);
        byte[] hrpBytes = Utils.copyAndFreeVec(hrpVecPointer[0], error);

        String hrp = null;
        try {
            hrp = new String(hrpBytes, "UTF-8");
        } catch (Exception e) {
            throw new SignerException("charset error");
        }

        return Pair.with(publicKey, hrp);

    }

}

File: oc/YeeSigner/YeeSigner/YeeSigner/Address.m
//
//  Address.m
//  YeeSigner
//
//  Created by GB on 2020/8/2.
//  Copyright © 2020 io.yeeco. All rights reserved.
//

#import "Address.h"
#import "libyee_signer.h"
#import "ErrorUtils.h"
#import "Utils.h"

@implementation Address

+ (NSString *) encode:(NSData* ) publicKey hrp:(NSString *) hrp error:(NSError **) error {
    
    unsigned int err = 0;
    
    NSData* hrpData = [hrp dataUsingEncoding:NSUTF8StringEncoding];
    unsigned int* vecPointer = yee_signer_address_encode(publicKey.bytes,  (unsigned int)publicKey.length, hrpData.bytes, (unsigned int)hrpData.length, &err);
    
    if(err > 0) {
        *error = [ErrorUtils error:err];
        return nil;
    }
    
    NSData *vec = [Utils copyAndFreeVec:vecPointer error:&err];
    
    if(err > 0) {
        *error = [ErrorUtils error:err];
        return nil;
    }
    
    NSString *address = [NSString stringWithUTF8String:[vec bytes]];
    
    return address;

}

+ (void) decode:(NSString* ) address publicKey:(NSData **) publicKey hrp:(NSString **) hrp error:(NSError **) error {
    
    unsigned int err = 0;
    
    unsigned int *publicKeyPointer = nil;
    unsigned int *hrpPointer = nil;
    
    NSData* addressData = [address dataUsingEncoding:NSUTF8StringEncoding];
    
    yee_signer_address_decode(addressData.bytes, (unsigned int)addressData.length, &publicKeyPointer, &hrpPointer, &err);
    
    if(err > 0) {
        *error = [ErrorUtils error:err];
        return;
    }
    
    NSData *publicKeyVec = [Utils copyAndFreeVec:publicKeyPointer error:&err];
    
    if(err > 0) {
        *error = [ErrorUtils error:err];
        return;
    }
    
    NSData *hrpVec = [Utils copyAndFreeVec:hrpPointer error:&err];
    
    if(err > 0) {
        *error = [ErrorUtils error:err];
        return;
    }
    
    NSString *hrpStr = [NSString stringWithUTF8String:[hrpVec bytes]];
    
    *publicKey = publicKeyVec;
    
    *hrp = hrpStr;
    
}

@end
